File: python/树/1457. 二叉树中的伪回文路径.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def pseudoPalindromicPaths (self, root: TreeNode) -> int:
        res = []
        def dfs(root,path):
            if not root:
                return
            path.append(root.val)
            if not root.left and not root.right:
                res.append(path[:])
            dfs(root.left,path)
            dfs(root.right,path)
            path.pop()
        dfs(root,[])
        # print(res)
        i = 0
        for r in res:
            if self.isFackPalindrome(r):
                i+=1
        return i

    def isFakePalindrome(self,path):
        from collections import defaultdict
        d = defaultdict(int)
        for i in path:
            d[i] +=1
        count = 0
        for v in d.values():
            if v & 1:
                count += 1
            if count == 2:
                return False
        return True

File: go/tree/1457. 二叉树中的伪回文路径.go
package tree

func pseudoPalindromicPaths(root *TreeNode) int {
	var dfs func(*TreeNode, []int)
	var res [][]int
	var path []int
	dfs = func(root *TreeNode, path []int) {
		if root == nil {
			return
		}
		path = append(path, root.Val)
		if root.Left == nil && root.Right == nil {
			tmp := make([]int, len(path))
			copy(tmp, path)
			res = append(res, tmp)
		}
		dfs(root.Left, path)
		dfs(root.Right, path)
		path = path[:len(path)-1]
	}
	dfs(root, path)
	r := 0
	for _, p := range res {
		if isFakePalindrome(p) {
			r++
		}
	}
	return r
}

func isFakePalindrome(path []int) bool {
	counter := make(map[int]int)
	for _, v := range path {
		counter[v]++
	}
	c := 0
	for _, v := range counter {
		if v%2 != 0 {
			c++
		}
		if c == 2 {
			return false
		}
	}
	return true
}
